### PR TITLE
feat(sdk): add GPU runtime resource request and fix spelling in runtime_resouce_request. Fixes  #4877. Fixes #1252

### LINKED ISSUE #1252: Cannot set the (set_gpu_limit) gpu num from user params
The set_gpu_limit only accept const number, cannot get input from user  parmas

### LINKED ISSUE #4877: Add in support for podSpecPatch
Often defining CPU/Mem/GPU resources at compile time is too inflexible. This is especially important if when you deploy your pipeline into different environments you may be sending through smaller datasets to expedite testing and want to recoup hardware costs as we no longer require such large machines.

To do this with Argo they have implemented `podSpecPatch` but this is not exposed in `kfp`, see their examples:
* https://github.com/argoproj/argo/blob/master/examples/pod-spec-from-previous-step.yaml
* https://github.com/argoproj/argo/blob/master/examples/pod-spec-patch.yaml
* https://github.com/argoproj/argo/blob/master/examples/pod-spec-yaml-patch.yaml
* https://github.com/argoproj/argo/blob/master/examples/pod-spec-patch-wf-tmpl.yaml

Not clear to me the exact syntax changes to the KFP DSL that'd be required to support this but from our perspective the per-template injection of resources is likely the most important, either from a [prior step](https://github.com/argoproj/argo/blob/master/examples/pod-spec-from-previous-step.yaml) or via a [pipeline parameter](https://github.com/argoproj/argo/blob/master/examples/pod-spec-patch.yaml).

/area sdk
/kind feature

### PR DESCRIPTION
**Pull request by NikeNano**
**Description of your changes:**
Fixes #4877 
Fixes #1252 

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: samples/core/resource_spec/runtime_resource_request.py [hunk 1]
```diff
@@ -23,26 +23,26 @@ def training_op(n: int) -> int:
...
 
 @components.create_component_from_func
-def generate_resouce_request() -> NamedTuple('output', [('memory', str), ('cpu', str)]):
+def generate_resouce_request() -> NamedTuple('output', [('memory', str), ('cpu', str), ('nbr_gpus', str), ('gpu_vendor', str)]):
     '''Returns the memory and cpu request'''
     from collections import namedtuple
     
-    resouce_output = namedtuple('output', ['memory', 'cpu'])
-    return resouce_output('500Mi', '200m')
+    resouce_output = namedtuple('output', ['memory', 'cpu', 'nbr_gpu', 'gpu_vendor'])
```
Comment (NikeNano) [inline samples/core/resource_spec/runtime_resource_request.py:30]:
```suggestion
    resource_output = namedtuple('output', ['memory', 'cpu', 'nbr_gpu', 'gpu_vendor'])
```

File: samples/core/resource_spec/runtime_resource_request.py [hunk 2]
```diff
@@ -23,26 +23,26 @@ def training_op(n: int) -> int:
...
+    return resouce_output('500Mi', '200m', 'nvidia', '1')
```
Comment (NikeNano) [inline samples/core/resource_spec/runtime_resource_request.py:31]:
```suggestion
    return resource_output('500Mi', '200m', 'nvidia', '1')
```

File: samples/core/resource_spec/runtime_resource_request_gpu.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+# limitations under the License.
+
+import kfp
+from kfp import dsl, components
+from typing import NamedTuple
+
+
+def training_job():
+    # quickly allocate a lot of memory to verify memory is enough
+    # image base_image='pytorch/pytorch:1.7.1-cuda11.0-cudnn8-runtime',
```
Comment (NikeNano) [inline samples/core/resource_spec/runtime_resource_request_gpu.py:22]: drop

File: samples/core/resource_spec/runtime_resource_request_gpu.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        packages_to_install=[]
+    )
+
+@components.create_component_from_func
+def generate_resouce_request() -> NamedTuple('output', [('nbr_gpus', str), ('gpu_vendor', str), ('constrain_type', str), ('constrain_value', str)]):
+    '''Returns the memory and cpu request'''
+    from collections import namedtuple
+    resouce_output = namedtuple('output', ['nbr_gpu', 'gpu_vendor', 'constrain_type', 'constrain_value'])
+
+    return resouce_output( 'nvidia', '1', 'cloud.google.com/gke-accelerator', 'nvidia-tesla-p4')
```
Comment (NikeNano) [inline samples/core/resource_spec/runtime_resource_request_gpu.py:42]:
```suggestion
    return resource_output( 'nvidia', '1', 'cloud.google.com/gke-accelerator', 'nvidia-tesla-p4')
```

File: samples/core/resource_spec/runtime_resource_request_gpu.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        training_job,
+        base_image='pytorch/pytorch:1.7.1-cuda11.0-cudnn8-runtime',
+        packages_to_install=[]
+    )
+
+@components.create_component_from_func
+def generate_resouce_request() -> NamedTuple('output', [('nbr_gpus', str), ('gpu_vendor', str), ('constrain_type', str), ('constrain_value', str)]):
+    '''Returns the memory and cpu request'''
+    from collections import namedtuple
+    resouce_output = namedtuple('output', ['nbr_gpu', 'gpu_vendor', 'constrain_type', 'constrain_value'])
```
Comment (NikeNano) [inline samples/core/resource_spec/runtime_resource_request_gpu.py:40]:
```suggestion
    resource_output = namedtuple('output', ['nbr_gpu', 'gpu_vendor', 'constrain_type', 'constrain_value'])
```

File: sdk/python/kfp/compiler/_op_to_template.py [hunk 6]
```diff
@@ -290,10 +302,10 @@ def _op_to_template(op: BaseOp):
 
     # Runtime resource requests
     if isinstance(op, dsl.ContainerOp) and ('resources' in op.container.keys()):
-        podSpecPatch = {}
         for setting, val in op.container['resources'].items():
             for resource, param in val.items():
-                if (resource in ['cpu', 'memory']) and re.match('^{{inputs.parameters.*}}$', param):
+                print(f"The resource is : {resource}, {type(resource)} and the item is: {param}, {type(param)}")
```
Comment (NikeNano) [inline sdk/python/kfp/compiler/_op_to_template.py:307]:
```suggestion
```

File: sdk/python/kfp/dsl/_container_op.py [hunk 7]
```diff
@@ -368,21 +368,28 @@ def set_gpu_limit(self, gpu, vendor='nvidia') -> 'Container':
...
+      vendor(Union[str, PipelineParam]): Optional. A string which is the vendor of the requested gpu.
         The supported values are: 'nvidia' (default), and 'amd'. The value is
         ignored in v2.
     """
-    self._validate_positive_number(gpu, 'gpu')
 
-    if self._container_spec:
-      # For backforward compatibiliy, allow `gpu` to be a string.
-      self._container_spec.resources.accelerator.count = int(gpu)
+    ## Should be and or here
```
Comment (NikeNano) [inline sdk/python/kfp/dsl/_container_op.py:377]:
```suggestion
```

**Comment by NikeNano** [comment]
@chensun could you take a look on this pr? I am still get issues with permissions to the runs @Bobgy:

```
https://4e18c21c9d33d20f-dot-datalab-vm-staging.googleusercontent.com/#/runs/details/203d6a9e-9749-4261-a202-b39da4616d66
```

gives me:

```
Google

403. That’s an error.

That’s all we know.
```

Thanks.

File: samples/core/resource_spec/runtime_resource_request_gpu.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+training_comp = components.create_component_from_func(
+        training_job,
+        base_image='pytorch/pytorch:1.7.1-cuda11.0-cudnn8-runtime',
+        packages_to_install=[]
+    )
+
+@components.create_component_from_func
+def generate_resource_request() -> NamedTuple('output', [('gpu_vendor', str), ('nbr_gpus', str), ('constrain_type', str), ('constrain_value', str)]):
+    '''Returns the gpu resource settings'''
```
Comment (chensun) [inline samples/core/resource_spec/runtime_resource_request_gpu.py:36]: nit: Use double quotes for docstring, and end the description with a period.

https://google.github.io/styleguide/pyguide.html#381-docstrings

File: samples/core/resource_spec/runtime_resource_request_gpu.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    from collections import namedtuple
+    resource_output = namedtuple('output', ['gpu_vendor', 'nbr_gpu', 'constrain_type', 'constrain_value'])
```
Comment (chensun) [inline samples/core/resource_spec/runtime_resource_request_gpu.py:38]: I was under the impression that the first argument to `namedtuple` needs to be the same as the left-hand side of the assignment. 
Maybe that's not a must-have but a convention?

**Comment by chensun** [comment]
> https://4e18c21c9d33d20f-dot-datalab-vm-staging.googleusercontent.com/#/runs/details/203d6a9e-9749-4261-a202-b39da4616d66

Failed to retrieve pod logs.
Maybe because it's stale, let's rerun the test.

/test kubeflow-pipelines-samples-v2

File: samples/core/resource_spec/runtime_resource_request_gpu.py (hunk 9 above)
Comment (NikeNano) [reply samples/core/resource_spec/runtime_resource_request_gpu.py:38]: It is not a requirement, https://docs.python.org/3/library/collections.html#collections.namedtuple. Example: 

```
from collections import namedtuple
Point = namedtuple('Tmp', ['x', 'y'])
p = Point(11, y=22)
p[0] + p[1]
```

>Maybe that's not a must-have but a convention?

Yes, it seems to be the convention, will update it.

**Comment by NikeNano** [comment]
> > https://4e18c21c9d33d20f-dot-datalab-vm-staging.googleusercontent.com/#/runs/details/203d6a9e-9749-4261-a202-b39da4616d66
> 
> Failed to retrieve pod logs.
> Maybe because it's stale, let's rerun the test.
> 
> /test kubeflow-pipelines-samples-v2

Still have the access problems so will have to wait for you to check it.

File: samples/core/resource_spec/runtime_resource_request_gpu.py (hunk 9 above)
Comment (chensun) [reply samples/core/resource_spec/runtime_resource_request_gpu.py]: Found the following from the doc you referenced.

> To support pickling, the named tuple class should be assigned to a variable that matches typename.

So it's bit more than just a convention.  :)

Comment (NikeNano) [reply samples/core/resource_spec/runtime_resource_request_gpu.py]: > So it's bit more than just a convention. :)

Good catch!

**Comment by chensun** [comment]
> @NikeNano: The following test **failed**, say `/retest` to rerun all failed tests:
> 
> Test name	Commit	Details	Rerun command
> kubeflow-pipelines-samples-v2	[7ef1ecf](https://github.com/kubeflow/pipelines/commit/7ef1ecff0a1a7755ec9727c217e6cfde1edec8f2)	[link](https://oss-prow.knative.dev/view/gs/oss-prow/pr-logs/pull/kubeflow_pipelines/5972/kubeflow-pipelines-samples-v2/1413401083144835072)	`/test kubeflow-pipelines-samples-v2`

The error was

> This step is in Pending state with this message: Unschedulable: 0/3 nodes are available: 3 Insufficient nvidia.com/gpu.

My guess is our test cluster was not set up with GPU enabled? @Bobgy, do you know?

**Comment by NikeNano** [comment]
> This step is in Pending state with this message: Unschedulable: 0/3 nodes are available: 3 Insufficient nvidia.com/gpu.

Sounds to me like we have GPU:s but all the resources(3 gpus) are used when the test. I think it would be fine to remove this tests as an alternative as well @chensun, however it would be awesome to test it first, dont have access to and GPU resource to try it out on though.

**Comment by chensun** [comment]
> > This step is in Pending state with this message: Unschedulable: 0/3 nodes are available: 3 Insufficient nvidia.com/gpu.
> 
> Sounds to me like we have GPU:s but all the resources(3 gpus) are used by the test suite. I think it would be fine to remove this tests as an alternative as well @chensun, however it would be awesome to test it first, dont have access to and GPU resource to try it out on though.

Yes, I think you're right. If you could test this manually in your local environment, I agree we probably don't need sample (e2e) test for this feature. Unit tests to check compilation result should be just enough.
Thanks!

**Comment by NikeNano** [comment]
Could you take a look again @chensun.

File: samples/core/resource_spec/runtime_resource_request_gpu.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+@dsl.pipeline(
+    name='Runtime resource request pipeline',
+    description='An example on how to make resource requests at runtime.'
+)
+def resource_request_pipeline():
+    resource_task = generate_resource_request()
+
+    traning_task = training_comp().set_gpu_limit(resource_task.outputs['nbr_gpus'], resource_task.outputs['gpu_vendor'])\
+        .add_node_selector_constraint(resource_task.outputs['constrain_type'], resource_task.outputs['constrain_value'])
```
Comment (Bobgy) [inline samples/core/resource_spec/runtime_resource_request_gpu.py:50]: nit: I think these are called "constraints"

File: sdk/python/tests/compiler/compiler_tests.py [hunk 11]
```diff
@@ -419,6 +419,36 @@ def my_pipeline(memory: str, cpu: str):
 
       self.assertEqual(template['podSpecPatch'], '{"containers": [{"name": "main", "resources": {"requests": {"cpu": "{{inputs.parameters.memory}}"}}}]}')
       
+  def test_py_runtime_gpu_request(self):
+      """Test GPU request."""
+
+      def my_pipeline(nbr_gpus: int, gpu_vendor: str):
+        some_op().set_gpu_limit(nbr_gpus, gpu_vendor)
+      
+      # .add_node_selector_constraint(resource_task.outputs['constrain_type'], resource_task.outputs['constrain_value'])
```
Comment (chensun) [inline sdk/python/tests/compiler/compiler_tests.py:428]: nit: remove? (same for down below)

**Review by chensun — COMMENTED** [review]
/lgtm

File: samples/core/resource_spec/runtime_resource_request_gpu.py (hunk 10 above)
Comment (NikeNano) [reply samples/core/resource_spec/runtime_resource_request_gpu.py:50]: I think this is a combination, for the gpu limits I think it is referred to in the docs under [`resource`](https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/) and node constraints as `constraints`. Could update to name it `resource_constraints_task` to be complete though.

File: sdk/python/tests/compiler/compiler_tests.py (hunk 11 above)
Comment (NikeNano) [reply sdk/python/tests/compiler/compiler_tests.py]: fixed.

File: samples/core/resource_spec/runtime_resource_request_gpu.py (hunk 10 above)
Comment (Bobgy) [reply samples/core/resource_spec/runtime_resource_request_gpu.py]: Nevermind, this is only a sample. We don't need to check all the naming strictly.

I think it's good enough!

**Comment by Bobgy** [comment]
/lgtm
/approve
@NikeNano thank you for adding the highly wanted feature!
@chensun thank you for the review!

**Comment by shashisingh** [comment]
@NikeNano , I am using kfp python sdk version 1.8.0.

pip freeze | grep kfp
kfp==1.8.0
kfp-pipeline-spec==0.1.10
kfp-server-api==1.6.0

When I check the YAML generated after running "python   samples/core/resource_spec/runtime_resource_request_gpu.py", I see that nodeSelector is not being set correctly -> nodeSelector: {}.  Run time gpu settings seem to be working fine and I can see required podSpecPatch in the YAML file. I have attached generated YAML file. Thanks.

[runtime_resource_request_gpu.py.yaml.txt](https://github.com/kubeflow/pipelines/files/7119248/runtime_resource_request_gpu.py.yaml.txt)

**Comment by NikeNano** [comment]
> @NikeNano , I am using kfp python sdk version 1.8.0.
> 
> pip freeze | grep kfp
> kfp==1.8.0
> kfp-pipeline-spec==0.1.10
> kfp-server-api==1.6.0
> 
> When I check the YAML generated after running "python samples/core/resource_spec/runtime_resource_request_gpu.py", I see that nodeSelector is not being set correctly -> nodeSelector: {}. Run time gpu settings seem to be working fine and I can see required podSpecPatch in the YAML file. I have attached generated YAML file. Thanks.
> 
> [runtime_resource_request_gpu.py.yaml.txt](https://github.com/kubeflow/pipelines/files/7119248/runtime_resource_request_gpu.py.yaml.txt)

Thank you for reporting this @shashisingh,checked the code and there is a bug here: https://github.com/kubeflow/pipelines/blob/19b3de4dd049052de74868ad0dfc931a4c700265/sdk/python/kfp/compiler/_op_to_template.py#L321 where the `podSpecPath` is overwritten. I will make a PR during the day with the fix: 

```
if not 'containers' in podSpecPatch:
    podSpecPatch["containers"] = [{'name':'main', 'resources':{}}]
```

FYI: it is normally better to create issues like this as separate `issues` so we have the correct tracking of them.

**Comment by shashisingh** [comment]
@NikeNano I will create an issue for you to track.

**Pull request merged**